pkg/gardenlet/controller/backupentry/backup_entry_migration_controller.go

```diff
@@ -0,0 +1,198 @@
+			return reconcile.Result{}, fmt.Errorf("could not clear migration start time: %w", err)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Set the migration start time if needed
+	if backupEntry.Status.MigrationStartTime == nil {
+		log.Debugf("[BACKUPENTRY MIGRATION] Setting migration start time")
+		if err := setMigrationStartTime(ctx, gardenClient.Client(), backupEntry, &metav1.Time{Time: time.Now().UTC()}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not set migration start time: %w", err)
+		}
+	}
+
+	// If the restore annotation is set or the grace period is elapsed and migration is not currently in progress,
+	// update the backup entry status to force the restoration (fallback to the "bad case" scenario)
+	log.Debugf("[BACKUPENTRY MIGRATION] Checking if the backup entry should be forcefully restored")
+	if hasForceRestoreAnnotation(backupEntry) || r.isGracePeriodElapsed(backupEntry) && !r.isMigrationInProgress(backupEntry) {
+
+		log.Infof("[BACKUPENTRY MIGRATION] Updating backup entry status to force restoration")
+		if err := updateStatusForRestore(ctx, gardenClient.Client(), backupEntry); err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not update backup entry status to force restoration: %w", err)
+		}
+
+		if hasForceRestoreAnnotation(backupEntry) {
+			log.Debugf("[BACKUPENTRY MIGRATION] Removing force-restore annotation")
+			if err := removeForceRestoreAnnotation(ctx, gardenClient.Client(), backupEntry); err != nil {
+				return reconcile.Result{}, fmt.Errorf("could not remove force-restore annotation: %w", err)
+			}
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// Requeue after the configured sync period
+	return reconcile.Result{RequeueAfter: r.config.Controllers.BackupEntryMigration.SyncPeriod.Duration}, nil
+}
+
+func (r *migrationReconciler) isGracePeriodElapsed(backupEntry *gardencorev1beta1.BackupEntry) bool {
+	r.logger.Infof("%+v", backupEntry.Status)
+	r.logger.Infof("%+v", r.config.Controllers.BackupEntryMigration)
+	return time.Now().UTC().After(backupEntry.Status.MigrationStartTime.Add(r.config.Controllers.BackupEntryMigration.GracePeriod.Duration))
+}
+
+func (r *migrationReconciler) isMigrationInProgress(backupEntry *gardencorev1beta1.BackupEntry) bool {
+	staleCutoffTime := metav1.NewTime(time.Now().UTC().Add(-r.config.Controllers.BackupEntryMigration.LastOperationStaleDuration.Duration))
+	lastOperation := backupEntry.Status.LastOperation
+	return lastOperation != nil &&
+		lastOperation.Type == gardencorev1beta1.LastOperationTypeMigrate &&
+		(lastOperation.State == gardencorev1beta1.LastOperationStateProcessing || lastOperation.State == gardencorev1beta1.LastOperationStateError) &&
+		!lastOperation.LastUpdateTime.Before(&staleCutoffTime)
+}
+
+func setMigrationStartTime(ctx context.Context, c client.Client, backupEntry *gardencorev1beta1.BackupEntry, migrationStartTime *metav1.Time) error {
+	patch := client.StrategicMergeFrom(backupEntry.DeepCopy())
```

Comment:
Nit: `client.MergeFrom` would be sufficient here, right?

Reply:
OK